docs/architecture/adr-050-sign-mode-textual.md

```diff
@@ -129,6 +130,30 @@ Msg (<int>/<int>): <string>           // E.g. Msg (1/2): bank v1beta1 send coins
 End of transaction messages
 ```
 
+### 9. Custom `Msg`-Renderers
+
+Application developers may choose to not follow default value renderers' output for their own `Msg`s. In this case, they can implement their own custom `Msg`-renderer. This is similar to [EIP4430](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-4430.md), where the smart contract developer chooses the description string to be shown to the end user.
+
+This is done by setting the `cosmos.msg.v1.textual.custom_renderer` Protobuf option to `true`. This option MUST be set on a `sdk.Msg`, i.e. on a Protobuf message whose name starts with `Msg`.
+
+```proto
+message MsgFooBar {
+  option (cosmos.msg.v1.textual.custom_renderer) = true;
+
+  // proto fields
+}
+```
+
+When this option is set to true on a `Msg`, the implementation CANNOT use the default value renderers to render the `Msg`. Instead, it MUST check that a custom `Msg`-renderer is implemented for `MsgFooBar` (e.g. registered in a registry), or else the implementation MUST throw an error or panic.
+
+Moreover, the implementation must provide 2 functions: one for formatting from Protobuf to string, and one for parsing string to Protobuf. These 2 functions are provided by the application devloper. To satisfy point #1, these 2 functions MUST be bijective with each other as inverse.
```

Comment:
I'd add that:

- The two data types being interconverted are a `Msg` using proto equality and an array of strings using Go's deep equality.
- Both functions accept the current context as an argument, allowing stored state to be read, e.g. account-specific settings which might affect the translations.
- Both functions SHOULD NOT modify state.
- Note that if state is modified in a way that affects the translation between the signature being created and the signature being verified, the transaction will be rejected.

Reply:
Instead of:

> Both functions SHOULD NOT modify state.

I would say: renderer functions must be pure (must not have any side effect such us state modification or global parameter update).

Reply:
Also we may be more clear that "for every `Msg m1, m2: m1.Equals(m2) <==> renderer.Format(m1) == renderer.Format(m2)`

Reply:
> I would say: renderer functions must be pure

Renderer functions should be able to read from state though, which make them unpure. To answer your question https://github.com/cosmos/cosmos-sdk/pull/12785#discussion_r936435378, it's why we pass the `ctx` too.

Reply:
What's the use case for render function to read additional data from the state?

Reply:
For example a mapping in state from address to username/nickname. Then TEXTUAL would shown "mom" or "john" instead of "agoric1abcdef...."

Reply:
Do we really want that? Wallet, to prepare that message will need to have an access to the latest state. Ideally, wallet should be able to create and sign transaction fully offline (eg by tracking locally the nonce and account number).

Reply:
Currently, even without custom renderers, wallets already need access to state for coin metadata. So I don't think it's a big addition here.

Reply:
It's not needed to create and sign transaction. Only account nonce and number is needed - but they can track it by themselves and sync once for a while, or on request.

Reply:
The custom renderer needs access to state at runTx time so it can generate the same petname substitutions.

Reply:
> For example a mapping in state from address to username/nickname. Then TEXTUAL would shown "mom" or "john" instead of "agoric1abcdef...."

https://docs.microsoft.com/en-us/dotnet/standard/security/cryptographic-signatures

> Cryptographic digital signatures use public key algorithms to provide data integrity. When you sign data with a digital signature, someone else can verify the signature, and can prove that the data originated from you and was not altered after you signed it. 

If the signature of a message is based on anything other than the message itself, then the signature doesn't actually provide data integrity, and the signature isn't actually a signature.

> Renderer functions should be able to read from state though, which make them unpure.

Assuming renderer functions produce signable bytes, it is _essential_ that they are pure.

Reply:
> For example a mapping in state from address to username/nickname. Then TEXTUAL would shown "mom" or "john" instead of "agoric1abcdef...."

This requires that username/nickname mappings are one-to-one with account addresses. Is that guaranteed by whatever system you'd be querying? If you could alias both `agoric1=mom` and `hacker1=mom` then transactions `from=me to=agoric1 amount=100` and `from=me to=hacker1 amount=100` would produce the same signable bytes. I hope the problems there are clear 😉

Reply:
I also think offline signing is worth maintaining. The more state is required, the harder that gets. Denom metadata can easily be shipped to the user in a config as this hardly changes. A message renderer that accesses state was requested by a user but is not what I wish to build and something that should remain opt-in.

Regarding the bijectivity: I'm not convinced we use this property. The way we do it right now with Amino JSON is: create proto document, serialize document to proto bytes (A), generate Amino JSON from document (B). Now B is signed and A is submitted to chain. We don't need Amino JSON -> proto as part of the signature verification or signing (correct me if I'm wrong). Since A and B are different, it is perfectly possible to have a Amino JSON converter that changes meaning if implemented that way.

Reply:
>  Since A and B are different, it is perfectly possible to have a Amino JSON converter that changes meaning if implemented that way.

Yes, and I hope it's clear why this is a problem.

Reply:
What about this case: I sign `from=me to=agoric1 amount=100` assuming `agoric1=ABC`. But in the meantime the state changes to `agoric1=XYZ`. Now the chain made me sign something I did not want to sign.

Reply:
> Yes, and I hope it's clear why this is a problem.

Assuming I have to trust the chain's code anyways, it is actully not that clear to me how it becomes a problem.

Reply:
I was just going to write _There is one more problem with the state: it can change_. And before hitting send, I saw @webmaster128 comment.

There could be even more traps if we enable state access.

Reply:
> Assuming I have to trust the chain's code anyways, it is actully not that clear to me how it becomes a problem.

...what?

You have to trust the chain's code to an extent. But message signing is one of many capabilities that decentralized/permissionless networks employ so that users _don't need_ to trust the chain, its implementation, or its participants.

The point of signing a message is to guarantee, in a cryptographic sense, the integrity of a specific sequence of bytes. That integrity is verifiable and trustable by anyone, no matter what intermediating users want to do. But only if you play by the rules! The rules say that, given a privkey/pubkey pair, signing and verifying look like this:

```
fn sign(msg, privkey) -> signature
fn verify(msg, signature, pubkey) -> bool
```

You sign a specific msg. A msg is a specific sequence of bytes. Not a domain object, not one of N possible renderings of a sequence of bytes, and so on. A signature for some msg M establishes the integrity of that msg M _only_.

If the Amino rendering of a msg transforms one of its fields in one way, and a textual rendering transforms that field (or a different field, or no fields) in a different way, then the renderings yield different byte sequences. So you could produce two different msgs that would have the same rendered form. If you swapped actual msg M1 out for a different msg M2 that rendered the same way, and then asked the signer to sign M2, you could then apply that signature to M1, and it would parse as valid. But the signer didn't sign M1, they signed M2. So you've lost integrity.

The signature of a msg/tx has to be on its canonical Merkelized bytes. You can't sign a rendering.

Reply:
> You have to sign the canonical, Merkelized bytes

The merklelized tx bytes include the tx signature, so by definition you can't sign the whole merklelized tx. What you sign contains the tx minus the signature.

> Signing a rendering doesn't provide integrity.

I argue that as long as `render_foo()` is 1/ part of consensus (so publicly and deterministically reproduceable by anyone who wants to verify sigs), and 2/ guarantees losslessness and non-malleability, then it provides integrity.

In all the SDK's sign modes, we sign a rendering of the tx (called `SignDoc*`).
- DIRECT: we sign the proto bytes of [this guy](https://github.com/cosmos/cosmos-sdk/blob/daca54d933d4a7c736324cc4add9f74d5394b648/proto/cosmos/tx/v1beta1/tx.proto#L48). Probably the closest to what's merklelized on chain, but still not quite that
- DIRECT_AUX: proto bytes of [this guy](https://github.com/cosmos/cosmos-sdk/blob/daca54d933d4a7c736324cc4add9f74d5394b648/proto/cosmos/tx/v1beta1/tx.proto#L71)
- AMINO_JSON: we sign the amino JSON representation of the tx

I'd go one step further than Jim's https://github.com/cosmos/cosmos-sdk/pull/12785#discussion_r943025204 to say that your concerns touch on existing sign modes too (especially the 2 latter). If in your eyes this is critical, SEV-0, red-alert, consider reporting this as a security issue, so that the relevant eyes takes a look at this quickly.

Reply:
I think we should be signing the raw txbody and authinfo bytes in sign mode textual. They can be "advanced mode" strings, but if they're not there the tx is malleable

Reply:
After some more thinking, I came around on this, and I agree with @aaronc we should sign the hash of [body bytes](https://github.com/cosmos/cosmos-sdk/blob/daca54d933d4a7c736324cc4add9f74d5394b648/proto/cosmos/tx/v1beta1/tx.proto#L52) and [auth_info_bytes](https://github.com/cosmos/cosmos-sdk/blob/daca54d933d4a7c736324cc4add9f74d5394b648/proto/cosmos/tx/v1beta1/tx.proto#L56) as they are included on-chain. Or else we would have tx hash malleability (though the tx values are not malleable thanks to bijectivity).

I'll add a paragraph on this.

Reply:
> I argue that as long as render_foo() is 1/ part of consensus (so publicly and deterministically reproduceable by anyone who wants to verify sigs), and 2/ guarantees losslessness and non-malleability, then it provides integrity.

Let's say we have a renderer that transforms addresses to nicknames, using on-chain "address books" which users can define for themselves, and which don't permit duplicate names. The renderer is part of consensus, the transformation is lossless and non-malleable. All points satisfied.

If user `agoric1` has addrbook `{agoric1=me agoric2=alice}` then the tx `{from=agoric1 to=agoric2 amount=100}` renders as `{from=me to=alice amount=100}` and produces some signature S1.

A malicious user `agoricX` can observe that tx and its rendering to create addrbook `{agoric1=me agoricX=alice}` and then tx `{from=agoric1 to=agoricX amount=100}` which would render as `{from=me to=alice amount=100}` and allow the malicious user to use the observed signature S1. I believe signature verification of this malicious tx would succeed?

This is just an example. Maybe other tx/msg properties would prevent this specific scenario. Maybe you could define "rules" for renderers like "don't transform sender addresses" or something. But I hope it's clear that the underlying issue remains. Signatures provide integrity only of the actual bytes signed, any transformation breaks the trust model.

Reply:
I added c3ac6edca6765e21f141d334bded76895f25dbfa, LMK (maybe in a new github thread) if that addresses your concerns.

Reply:
[The example of the addressbook is back on-topic. :-)] The custom renderer must know to use the addressbook associated with the signer/sender for this transaction, so `agoricX` can construct whatever addressbook they'd like, but the rendering would use the addressbook of `agoric1`. It's not too much of a burden to expect the custom renderer to get this right - we already expect Msg implementors to implement `GetSigners()` correctly.

Reply:
> The custom renderer must know to use the addressbook associated with the signer/sender for this transaction, so agoricX can construct whatever addressbook they'd like, but the rendering would use the addressbook of agoric1. It's not too much of a burden to expect the custom renderer to get this right - we already expect Msg implementors to implement GetSigners() correctly.

If msg implementors don't implement GetSigners correctly, I assume it will produce an error, or halt the chain, or in some way refuse to accept the bad data? If so, that's fine! But I've presented a categorically different situation. If msg implementors accidentally take the wrong addrbook from state, then it's possible for malicious users to fake signatures, and malicious data to enter the chain.

Reply:
> The signature must be over data that includes the text presented to the user, otherwise malicious wallet software could send the wrong text to the Ledger and get the user to sign a completely different transaction.

There is an unavoidable trust assumption in any human-readable information derived from non-human-readable data. The text is always going to be model of the tx, not the tx itself. Cryptographic signatures operate on data, they aren't meant to be human-auditable, right? If this is something that the maintainers believe to be a requirement delivered by e.g. Ledger, then there must have been some amount of miscommunication or misunderstanding. It just don't make sense.

Reply:
No - if, for instance, the bank "send" message messed up `GetSigners()` and specified that the recipient needed to sign rather than the sender, then the chain would happily let users drain each other's accounts. This method absolutely must be implemented correctly for the integrity of the chain. That's why I say that it's not a new burden to expect the renderer to select the right user's addressbook, and a screwup in doing so is not a new class of error.

Reply:
The whole reason the Ledger has a screen is so that the user can understand what they are signing, and that only makes sense if the presented text cannot be manipulated without invalidating the signature.

Reply:
> The whole reason the Ledger has a screen is so that the user can understand what they are signing, and that only makes sense if the presented text cannot be manipulated without invalidating the signature.

Cosmos transactions cannot be rendered in a way that's simultaneously human-readable, lossless, and non-malleable.

Reply:
> Cosmos transactions cannot be rendered in a way that's simultaneously human-readable, lossless, and non-malleable.

Bummer. Someone better explain that to SIGN_MODE_LEGACY_AMINO_JSON.

Reply:
JSON is arguably human-readable — it is for me! 😉 — but it is both malleable and lossy. At least, per the spec. if that sign mode uses a lossless and deterministic JSON encoder, then it may be OK!

But I did some research. SIGN_MODE_LEGACY_AMINO_JSON apparently (?) uses [tendermint/go-amino](https://github.com/tendermint/go-amino) to encode and decode, and I guess specifically the JSON methods in that package. I believe execution eventually gets to [Codec.encodeReflectJSON](https://github.com/tendermint/go-amino/blob/master/json-encode.go#L23), which forwards e.g. map encoding over to [encodeReflectJSONMap](https://github.com/tendermint/go-amino/blob/8e779b71f40d175cd1302d3cd41a75b005225a7a/json-encode.go#L343). That function iterates over the map using [reflect.Value.MapKeys](https://github.com/tendermint/go-amino/blob/8e779b71f40d175cd1302d3cd41a75b005225a7a/json-encode.go#L370), which [per documentation] (https://pkg.go.dev/reflect#Value.MapKeys)

> MapKeys returns a slice containing all the keys present in the map, **in unspecified order**. 

Can valid txs/msgs contain maps? I'm not sure, actually! But if they can, then not only is this sign mode lossy and malleable, it is also non-deterministic! :o

Reply:
Resolving this thread as off-topic. There's a security thread going on.